core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2237,7 +2209,10 @@ class Log(@volatile private var _dir: File,
     def deleteSegments(): Unit = {
       info(s"Deleting segments ${segments.mkString(",")}")
       maybeHandleIOException(s"Error while deleting segments for $topicPartition in dir ${dir.getParent}") {
-        segments.foreach(_.deleteIfExists())
+        segments.foreach { segment =>
+          segment.deleteIfExists()
+          producerStateManager.deleteIfExists(segment.baseOffset)
```

Comment:
Hmm, this can be a bit tricky. When we replace old segments with a new segment in LogCleaner, each of the old segment will be deleted. However, the first old segment has the same offset as the new segment. So, we don't want to just delete the producer snapshot corresponding to the first old segment.